app/controllers/users/sessions_controller.rb

```diff
@@ -31,15 +31,21 @@ def new
 
     def create
       session[:sign_in_flow] = :sign_in
-      return process_locked_out_session if session_bad_password_count_max_exceeded?
-      return process_locked_out_user if current_user && user_locked_out?(current_user)
+
+      failure_reason = nil
+      failure_reason = process_locked_out_session if session_bad_password_count_max_exceeded?
+      failure_reason = process_locked_out_user if current_user && user_locked_out?(current_user)
+      failure_reason = process_failed_captcha if !valid_captcha_result?
+      return if failure_reason
```

Comment:
Looks like we have multiple side effects we could perform, but only keep the last failure reason. Should we make failure reason an array/set/hash here and add to it?

Reply:
Oh thats a good idea.

Reply:
I think if I were to refactor this, I'd probably make this use our form submission convention, and log the failure reasons as the `FormResponse#error_details` instead. So I'm not sure it's too worthwhile to invest in `failure_reason`.

But one thing your comment made me realize is that these revisions unintentionally changed the priority of how failure reasons are selected in the case of multiple failures, since it now processes all of the reasons instead of bailing at the first one. I'll at least try to clean that up, and hopefully maintain more how it works currently.

Reply:
I updated this in a49374ef31 which results in a much smaller diff and avoids juggling `failure_reason`.